### PR TITLE
KEM Traits

### PR DESCRIPTION
**Pull request by rozbb**
Going off our [previous discussion](https://github.com/RustCrypto/traits/issues/498), I took a crack at implementing some KEM traits.

Sample usage, from the HPKE unit test:

```rust
// Make a sender and recipient keypair. gen_keypair is completely separate from the KEM traits
let (sk_sender, pk_sender) = gen_keypair(&mut rng);
let (sk_recip, pk_recip) = gen_keypair(&mut rng);

// Try an unauthed encap first. Check that the derived shared secrets are equal
let encapper = X25519Encap;
let (ek, ss1) = encapper.try_encap(&mut rng, &pk_recip).unwrap();
let ss2 = sk_recip.try_decap(&ek).unwrap();
assert_eq!(ss1, ss2);

// Now do an authenticated encap
let encapper = X25519AuthEncap(sk_sender, pk_sender.clone());
let (ek, ss1) = encapper.try_encap(&mut rng, &pk_recip).unwrap();
let ss2 = sk_recip.try_auth_decap(&ek, &pk_sender).unwrap();
assert_eq!(ss1, ss2);

// Now do an invalid authenticated encap, where the sender uses the wrong private key. This
// should produce unequal shared secrets.
let (rand_sk, _) = gen_keypair(&mut rng);
let encapper = X25519AuthEncap(rand_sk, pk_sender.clone());
let (ek, ss1) = encapper.try_encap(&mut rng, &pk_recip).unwrap();
let ss2 = sk_recip.try_auth_decap(&ek, &pk_sender).unwrap();
assert_ne!(ss1, ss2);

```

## Notes

* I made `EncappedKey` hold other KEM data in the form of associated types. I don't know of any cases where an encapsulated key type could and should be compatible with multiple pubkey or shared-secret types, so I think this is OK. It also makes the interface a lot cleaner.
* ~~I wrote `Serializable` and `Deserializable` traits because `PublicKey: AsRef<[u8]>` is not generally true (e.g., when `PublicKey` is a compressed curve point and its serialized format must be uncompressed). For the same reason, I require `EncappedKey`s to be serializable.~~ I actually disagree with this move. I reverted this and made `EncappedKey` a bag of bytes again. I also removed all trait bounds on `PublicKey`. We might want to make some guarantees about `PublicKey` though. Open to ideas.
* I dropped `AuthEncapsulator` altogether. I realized that you can just define `sk.try_encap()` to do the authenticated encapsulation. Thus, there's just 1 `Encapsulator` trait.
* In the same vein, you could also drop `AuthDecapsulator` by just folding the sender pubkey into the decapsulator itself. I don't think this makes much sense, though, since passing around a `Decapsulator` that only works for a single sender doesn't seem like a reasonable way to use a KEM interface.
*  I haven't written the infallible `encap()`, `decap()`, and `decap_auth()` methods yet. This is simple though.
* Besides the HPKE test I plan on writing a test using the KEMs in `pqcrypto`. Eventually, those impls will be added to `pqcrypto-compat`.

### CODE REVIEW

**Comment by rozbb** [comment]
One issue right now is that `EncappedKey::PublicKey` is supposed to represent both the recipient's public key and the sender's public key used for authentication. These are not the same key in general.

For example, in X3DH, the "public key" you encapsulate to is actually a tuple of an identity pubkey, a medium-term pubkey, and a signed one-time-prekey. The "public key" you use when auth-decapsulating is _just_ an identity pubkey.

A simple fix is to split `EncappedKey::PublicKey` into two types: `SenderPublicKey` and `RecipPublicKey`

**Comment by rozbb** [comment]
Ok I wrote 3 test implementations and I'm pretty happy with how it looks so far. I have it working for a very simple unauthenticated KEMs (Saber), a complicated authenticated KEMs (X3DH), and somewhere in between (HPKE's DHKEM)

File: kem/tests/saber.rs [hunk 1]
```diff
@@ -0,0 +1,75 @@
+#[cfg(test)]
+mod tests {
```
Comment (tarcieri) [inline kem/tests/saber.rs:2]: This is implicit for any tests in `tests/`

File: kem/tests/x3dh.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
+#[cfg(test)]
+mod tests {
```
Comment (tarcieri) [inline kem/tests/x3dh.rs:2]: Likewise here... this can be removed

File: kem/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+[![Build Status][build-image]][build-link]
+
+This crate contains traits which provide generic, object-safe APIs for
+establishing shared secrets using [a key encapsulation mechanism][1].
+
+[Documentation][docs-link]
+
+## Minimum Supported Rust Version
+
+Rust **1.41** or higher.
```
Comment (tarcieri) [inline kem/README.md:17]: Not sure what actual MSRV you have in mind, but this doesn't match what's in lib.rs, which is 1.56.

If you make this a 2021 edition crate, you can unify the toplevel rustdoc with what's in README.md, as it were.

**Comment by rozbb** [comment]
Thanks. Yeah those paragraphs were all stolen from the signature crate. Will fix

**Comment by rozbb** [comment]
I made the edition 2021 and standardized on whatever rustc version I'm running right now (1.56.1 2021-11-01). Is there a more principled way of picking an MSRV I should do?

**Comment by tarcieri** [comment]
We usually specify MSRV in terms of a minor version like 1.56 instead of 1.56.0. That works with the `rust-version` attribute as well.

In terms of selecting a version, Rust 2021 is MSRV 1.56 by definition, so that’s that.

File: kem/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+//! such a parameter for the return value, allowing it to be inferred by the type checker based on
+//! the desired KEM algorithm.
+//!
+//! The [`Signature`] trait is requires `as_bytes()` and `from_bytes()` methods, enforcing that
+//! encapsulated key types are thin wrappers around a "bag-of-bytes" serialization. Inspiration for
+//! this approach comes from the Ed25519 signature system, which was based on the observation that
+//! past systems were not prescriptive about how signatures should be represented on-the-wire, and
+//! that lead to a proliferation of different wire formats and confusion about which ones should be
+//! used. This crate aims to provide similar simplicity by minimizing the number of steps involved
+//! to obtain a serializable encapsulated key.
```
Comment (tarcieri) [inline kem/src/lib.rs:54]: You can use the following on Rust 1.56 to import the README.md into the rustdoc (and also doctest any examples in the README!)

```rust
#![doc = include_str!("../README.md")]
```

If you'd like to add additional toplevel rustdoc this syntax works:

```rust
#![doc = include_str!("../README.md")]

//! # Extra stuff here
//! 
//! This just shows up as additional concatenated Markdown
```

**Comment by rozbb** [comment]
I think I updated the deps in `Cargo.lock` beyond where they should've gone. This seems delicate. Should I just not include `Cargo.lock` in the diff at all?

Also separately I thought `Cargo.lock` wasn't really included in libs. It's not present in a few of the other RustCrypto libs I checked.

**Comment by tarcieri** [comment]
>  Should I just not include Cargo.lock in the diff at all?

I would suggest only including relevant changes, e.g. adding a new crate to the workspace.

> Also separately I thought Cargo.lock wasn't really included in libs. It's not present in a few of the other RustCrypto libs I checked.

This gets asked often enough we should probably write a FAQ.

While not checking in Cargo.lock for libraries is a common rule-of-thumb, it has resulted in some extremely hard-to-debug situations for us in the past.

Just as an example, we recently tried to debug a situation where Ed25519 signature verification was failing for an individual. Nobody else could reproduce it, but it was fully reproducible for that individual.

Turns out their Cargo.lock contained a release of the `sha2` crate with a miscomputation bug. Since we had patched the bug, anyone who had an older Cargo.lock with an earlier release didn't see the bug, and anyone who cloned the repo fresh and got the version with the bugfix didn't see the bug. But this perplexed several people for days.

Including Cargo.lock in the repo makes the build fully reproducible and eliminates these sorts of hard-to-debug scenarios.

**Comment by rozbb** [comment]
Ok, updated the lock file. I just copied the existing one and ran `cargo check`. I could use some help on the failing builds. Why is, e.g., `digest` failing to build because of the `edition = "2021"` key in `kem`?

**Comment by tarcieri** [comment]
Unfortunately if you have any 2021 edition crates in the workspace, the MSRV of the whole workspace becomes 1.56.

The workaround is to keep 2021 edition crates out of that workspace. Here's the hack I'm using with the `elliptic-curve` crate:

https://github.com/RustCrypto/traits/blob/master/elliptic-curve/Cargo.toml#L18-L20

Hopefully we can bump edition of all of the other crates in this repo in the next minor release (or in the case of `cipher`, maybe the release-after-next), but in the meantime, I think it's ok to copy this hack into `kem` as well.

**Comment by rozbb** [comment]
Should be good now

**Comment by tarcieri** [comment]
Looks like a promising start. Can you remove WIP? (or I can)

**Comment by tarcieri** [comment]
Thank you!

**Pull request merged**